web/page0/src/app/fetch.service.ts

```diff
@@ -118,5 +125,46 @@ export class FetchService {
     );
   }
 
-  constructor(private apiService: ApiService) {}
+  loadHosts(runtime: Runtime) {
+    if (runtime.status !== RuntimeStatus.valid || !runtime.zones) {
+      return of([]);
+    }
+
+    return forkJoin(
```

Comment:
instead of nested call, how about using merge?

hosts  = runtime.zones.flatmap(~~~);
envs = hosts.flatmap(...)

merge(hosts.map(make_message), env.map(make_message)).subscribe(this.store.dispatch)

return envs.scan or something

Reply:
Refactored.

Cannot subscribe inside of `loadHosts` since all the API calls should be canceled if RefreshService cancels the previous subscription. (i.e. the user presses the refresh button many times)

The resulting Observable of loadHosts is only used to mark completion of all API calls, thus annotated type as Observable<any>.